pandas/core/frame.py

```diff
@@ -2291,6 +2291,13 @@ def maybe_reorder(
             arr_columns = ensure_index(arr_columns)
             if columns is None:
                 columns = arr_columns
+
+                if index is not None:
+                    if isinstance(index, str) or not hasattr(index, "__iter__"):
+                        if index not in columns:
+                            result_index = [index]
+                            index = None
```

Comment:
I think that maybe is another bug, see https://github.com/pandas-dev/pandas/issues/47285#issuecomment-1150320114 where we get `NaN` in the index for a missing key.

Reply:
On main, this raises a KeyError right now, which is fine I think?

Reply:
i.e. if we do have data (not the empty list case) we should probably raise on missing keys

Reply:
Yep I think so too.

Reply:
> This should still raise, I think.

AFAICT it does still raise but with `ValueError: Length of values (4) does not match length of index (1)` instead?

on main it raises `KeyError: 'col_14'` iif the key is not present in any of the records, otherwise it creates an index with `NaN` for the missing keys.

Reply:
Yeah sorry this was misleading. I meant it should still raise a KeyError. This is way clearer than the other message

Reply:
> i.e. if we do have data (not the empty list case) we should probably raise on missing keys

Yeah agree with you. 
Furthermore, when we have EMPTY list and we pass some list index then currently on Main, it returns that list index as output with empty list. We want same behavior for string index as well.

Reply:
> This swallows KeyErrors when the iterable is populated but the column is not present.
> 
> ```
> data = [
>     {"col_1": 3, "col_2": "a"},
>     {"col_1": 2, "col_2": "b"},
>     {"col_1": 1, "col_2": "c"},
>     {"col_1": 0, "col_2": "d"},
> ]
> result = pd.DataFrame.from_records(data, index="col_14")
> print(result)
> ```
> 
> This should still raise, I think.
> 
> cc @simonjayhawkins

As @simonjayhawkins mentioned, and I also tested it, it generates ValueError: Length of values (4) does not match length of index (1) when we've iterable populated. And if it should raise KeyError then that is another issue.

Here, in my PR, I tried to address actual issue when we have empty list ([]) and we pass as string index, It gives KeyError. But Ideally in output we want Empty list along with passed string as index. 

We want Exact same behaviour when we have EMPTY list and we pass some list index then currently on Main, it returns that list index as output with empty list.

Code I change put, it will swallow KeyError only when we have EMPTY list passed with string index. Otherwise it won't get executed.

But please correct me if I misunderstood any of your concerns in discussion. Thanks.
![Screenshot from 2022-06-19 18-06-26](https://user-images.githubusercontent.com/30633049/174502133-44ca53b1-0978-4978-97da-f3da2b5c3bee.png)
![Screenshot from 2022-06-19 18-05-58](https://user-images.githubusercontent.com/30633049/174502134-b9bf0941-8374-44ae-ab69-33da023b67fb.png)

Reply:
My code example should still raise the same error. We do not want to change this behavior.

Reply:
You have to fix the bug without breaking anything else.

Please also add my example as a test with the KeyError as expected

Reply:
Ohk. I apologize, earlier I misunderstood it. I fixed it and now it is behaving as expected (exactly what currently in main this raises a KeyError). I have added your example as a test as well. Thank you.